pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java

```diff
@@ -338,11 +337,14 @@ public synchronized void removeConsumer(Consumer consumer, boolean isResetCursor
                     if (!isResetCursor) {
                         try {
                             topic.getManagedLedger().deleteCursor(cursor.getName());
+                            topic.getManagedLedger().removeWaitingCursor(cursor);
                         } catch (InterruptedException | ManagedLedgerException e) {
                             log.warn("[{}] [{}] Failed to remove non durable cursor", topic.getName(), subName, e);
                         }
                     }
                 });
+            } else {
+                topic.getManagedLedger().removeWaitingCursor(cursor);
```

Comment:
Great! It looks like this fixes #22157. I tested that all changes to addWaitingCursor can be removed that were made to fix #22191.

Reply:
yes, since this fix can fix #22157 and override #22191, we can keep #22191.

Reply:
sure, no need for explicit revert of #22191 with "git revert". I simply meant that changes to addWaitingCursor method can be removed.

Reply:
Oh, we can't remove this method. because we have removed the cursor from the waitingCursor  when consumer unsubscribe. and there are still race conditions with `consumer.close` and `checkForNewEntries`.which will lead to OOM
You can use testAddWaitingCursorsForNonDurable to reproduce

Reply:
I did test that. The test passes with this change: https://github.com/lhotari/pulsar/commit/d46ebc41

Reply:
That proves that the method could be removed, however since there's also `removeWaitingCursor`, it makes sense to have `addWaitingCursor` (which doesn't have any extra conditions).

Reply:
Thanks @Technoboy- , I got it reproduced with https://github.com/lhotari/pulsar/commit/e2efe24. 

Regardless of this, I think that the problem is deeper.
I was expecting that the race condition would be fixed by adding this to ManagedCursorImpl.checkForNewEntries method:
```
            if (isClosed()) {
                callback.readEntriesFailed(new CursorAlreadyClosedException("Cursor was already closed"), ctx);
                return;
            }
```
However, it doesn't work since when a cursor is deleted, it's status isn't switched to closed. This isn't very logical in the ManagedCursorImpl code and I think it should be fixed.

I made this commit in my experiment https://github.com/lhotari/pulsar/commit/43da5894 . It resolves the race condition that exists in `checkForNewEntries`.

Reply:
here's the diff of my experiment compared to the changes of this PR: https://github.com/Technoboy-/pulsar/compare/fix-22435...lhotari:pulsar:lh-remove-addWaitingCursor-conditions

I think that the proper way to fix the problem is to add the closed check to `checkForNewEntries` method and set the state to closed when the cursor is deleted.

Reply:
More better , addressed. Thanks @lhotari .